### PR TITLE
Move GMSL client types to Dendrite

### PR DESCRIPTION
**Pull request by kegsay**
GMSL is intended for Federation only. Sister PR to https://github.com/matrix-org/gomatrixserverlib/pull/357

### CODE REVIEW

File: syncapi/synctypes/filter.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+		Rooms:                   nil,
+		ContainsURL:             nil,
+	}
+}
+
+// DefaultRoomEventFilter returns the default room event filter used by the Matrix server if no
+// filter is provided in the request
+func DefaultRoomEventFilter() RoomEventFilter {
+	return RoomEventFilter{
+		Limit:       20,
```
Comment (devonh) [inline syncapi/synctypes/filter.go:138]: Why do we choose 20 here? Maybe worth a comment

File: syncapi/synctypes/filter.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	Types                   *[]string `json:"types,omitempty"`
+	LazyLoadMembers         bool      `json:"lazy_load_members,omitempty"`
+	IncludeRedundantMembers bool      `json:"include_redundant_members,omitempty"`
+	NotRooms                *[]string `json:"not_rooms,omitempty"`
+	Rooms                   *[]string `json:"rooms,omitempty"`
+	ContainsURL             *bool     `json:"contains_url,omitempty"`
+}
+
+// RoomEventFilter is used to define filtering rules for events in rooms
+type RoomEventFilter struct {
```
Comment (devonh) [inline syncapi/synctypes/filter.go:65]: Spec v1.4 adds unread_thread_notifications, can we trivially add that in here?

File: syncapi/synctypes/filter.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	Rooms        *[]string       `json:"rooms,omitempty"`
+	Ephemeral    RoomEventFilter `json:"ephemeral,omitempty"`
+	IncludeLeave bool            `json:"include_leave,omitempty"`
+	State        StateFilter     `json:"state,omitempty"`
+	Timeline     RoomEventFilter `json:"timeline,omitempty"`
+	AccountData  RoomEventFilter `json:"account_data,omitempty"`
+}
+
+// StateFilter is used to define filtering rules for state events
+type StateFilter struct {
```
Comment (devonh) [inline syncapi/synctypes/filter.go:52]: This is missing a limit field.

File: syncapi/synctypes/filter.go (hunk 2 above)
Comment (kegsay) [reply syncapi/synctypes/filter.go:65]: Yes we can.

File: syncapi/synctypes/filter.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+			Timeline:     DefaultRoomEventFilter(),
+		},
+	}
+}
+
+// DefaultEventFilter returns the default event filter used by the Matrix server if no filter is
+// provided in the request
+func DefaultEventFilter() EventFilter {
+	return EventFilter{
+		Limit:      20,
```
Comment (devonh) [inline syncapi/synctypes/filter.go:110]: Possible comment here for why 20

File: syncapi/synctypes/filter.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+// limitations under the License.
+
+package synctypes
+
+import (
+	"errors"
+)
+
+// Filter is used by clients to specify how the server should filter responses to e.g. sync requests
+// Specified by: https://matrix.org/docs/spec/client_server/r0.5.0.html#filtering
```
Comment (devonh) [inline syncapi/synctypes/filter.go:22]: Should we update this reference to the new spec location?
https://spec.matrix.org/v1.6/client-server-api/#filtering

**Review by devonh — APPROVED** [review]

**Pull request merged**